src/core/gpu_sw.cpp

```diff
@@ -89,6 +89,13 @@ void GPU_SW::Reset(bool clear_vram)
   m_backend.Reset(clear_vram);
 }
 
+void GPU_SW::UpdateResolutionScale()
```

Comment:
We should just remove this function entirely, or leave it empty. This setting is meant to change the uprender scale depending on the user's actual client window size. But this is not a safe operation on SW renderer because of its CPU requirements. (arguably we could allow for 1x and 2x switching, but not 4x, or something -- it's complicated so I decided to just avoid it instead for now).

If the GPU implementation is not changing the resolution scale, then this function is pointless.

OR -- since you're already removing most other aspects of the `use_resolution_scale` setting -- we could remove UpdateResolutionScale() entirely too.

`UpdateResolutionScale` is a feature ideal to HW GPU -- it auto-detects the best resolution scale for the given window size of the emulator. This is a good user-facing feature for the emulator running in GPU HW acceleration modes. **But DuckStation's implementation of it is needlessly complicated.** It should just be a simple per-frame check inside each HW GPU UpdateDisplay() instead of whatever it is now.

The amount of clutter around UpdateResolutionScale behavior on HW was actually the reason I decided to just make a separate setting for SW, because I didn't want to end up making all the changes to `use_resolution_scale` (but I only checked backend dependencies, I didn't consider if this had unwanted side effects in UI design).

At the point where you have removed most of that `use_resolution_scale` clutter anyway, might as well just go the extra step and also remove `UpdateResolutionScale` and just let backends manage it accordingly in the future. (and make it a setting! currently we can't even disable that behavior on HW GPU...)

Reply:
Oh I see, I didn't understand UpdateResolutionScale well then. I was trying to update the scale factor of the uprenderer on the fly when the user switched from 1x to 2x or 4x...

Reply:
OK, so let's flesh out the answer to this question:

1. should we bother to support at all DuckStation's current 'feature' of picking the best uprender scale for the user's given window size (an answer of NO means libretro sticks to the tried and true "you get the scale you pick" approach) ?
      - _(if YES, then we really must add an option to control/disable this behavior, which does not exist currently)_

Besides that, I still think it's best to have separate hw and sw resolution scale settings since they have different requirements and restrictions. So seems good to move forward with that.

Reply:
This is just my personal opinion, but I don't think this feature of Duckstation fits well in a libretro frontend. If it was just me, I would stick to the tried and true explicit setting.

> Besides that, I still think it's best to have separate hw and sw resolution scale settings since they have different requirements and restrictions. So seems good to move forward with that.
We usually avoid this, for example in mednafen, but in this particular case, I feel like it's OK to keep them separated for now.

What do you think about implementing the scale factor switch on the fly, without having to restart the core?
Other libretro core don't necessarily support this, some require a restart, some no.
If I remember well, the HW renderer supports that.

Reply:
It's trivial to support. Just need to do read out native VRAM and write it back as 2x or 4x. I'm already doing that for savestate restore. Same thing can be done on UpdateSettings().

Reply:
Regarding SW/HW scale option - We can have two options in DuckStation, and then have RetroArch UI just update both of them in unison, if RetroArch feels better having a single setting. This would be the most robust approach as it makes for a lean UI and would allow a user to still custom-configure things to some extent using settings.ini.

FWIW, a typical user flipping HW/SW actually wants/expects SW to automatically flip either into native res (max compatibility) or 2x res. I don't think anyone wants, or expects, SW to flip into 4x or higher, even though it's not at all uncommon to run GPU in 4x or sometimes even 8x mode on PS1. This is the main reason to track them independently.

Moreover, SW doesn't support higher than 4X unless I add a template for 8X or make a 'generic' template that supports any shift value (lowers perf for modes that use it). To complicate things further, HW GPU also supports integer scales such as 3x, 5x, 6x, etc with reasonably good compat. This is why I added some logic to GPU_SW to restrict the valid settings to 1,2,4 for now.

Reply:
> We can have two options in DuckStation, and then have RetroArch UI just update both of them in unison, if RetroArch feels better having a single setting. This would be the most robust approach as it makes for a lean UI and would allow a user to still custom-configure things to some extent using settings.ini.

^ I realize now this is impossible within the fully data driven retroarch UI API that's being used. There is only allowed a 1:1 correlation between text-editable settings and UI settings.

I could make a remapping but it's not worth the effort. Instead I'm going to use a single resolution_scale value and add a new option to disable that auto-size-by-window feature.

Reply:
OK, I realized the auto-resolution-sclae is already conditional, enabled only when GPU.resolution_scale=0. So no action needed.

I pieced together a simplified single-resolution-scale-option setting on my branch. We can close this PR now, I think.